### ISSUE TITLE
Target: T-TWR Plus

### ISSUE THREAD

**Issue by edgetriggered** [OpenRTX/OpenRTX#175]
### What

The purpose of this issue is to track high level features supported by the OpenRTX port to [LilyGo T-TWR Plus](https://www.lilygo.cc/products/t-twr-plus). Activity for this experimental work is taking place in the [ttwrplus](https://github.com/OpenRTX/OpenRTX/tree/ttwrplus) branch, while SA868S bandband firmware is tracked separately in the [sa8x8-fw](https://github.com/openrtx/sa8x8-fw) repository. Unlike other OpenRTX targets, which are ARM Cortex-M based and supported by Miosix, this platform uses the [Espressif ESP32-S3](https://www.espressif.com/en/products/socs/esp32-s3) system-on-chip with Xtensa LX7 cores.

### Who

The primary contacts for this port are:

- @n1zzo 
- @edgetriggered

Collaboration for this effort is taking place through the [OpenRTX Matrix Space](https://matrix.to/#/#openrtx:matrix.org) and/or [Discord](https://discord.gg/TbR2FVtMya), in the relevant T-TWR Plus room/channel.

### Resources
- [T-TWR Plus Hardware Documentation](https://openrtx.org/#/hardware/ttwrplus)
- [T-TWR Plus 2.0 Schematics](https://github.com/Xinyuan-LilyGO/T-TWR/blob/master/schematic/T-TWR-Plus_Rev2.0.pdf)
- [Zephyr API Documentation](https://docs.zephyrproject.org/apidoc/latest/index.html)

***

- [X] Preliminary OpenRTX application bootable with [Zephyr RTOS](https://github.com/zephyrproject-rtos/zephyr)
- [X] Power Management Unit supported through [XPowersLib](https://github.com/Xinyuan-LilyGO/T-TWR/tree/master/lib/XPowersLib)
- [X] Battery status
- [X] Buttons
- [X] Encoder wheel
- [X] Macro menu
- [X] FM transmit
- [X] FM receive
- [X] GPS receive
- [X] CTCSS support
- [X] Split frequencies (repeater support)
- [X] High transmission power (2.4W)
- [ ] APRS via BLE KISS interface (need a phone with aprs.fi app)
- [ ] Baseband signal routing through analog switches
- [ ] External microphone + speaker through soundcard daughterboard
- [ ] M17 transmit
- [ ] M17 receive
- [ ] Channel persistence

**Comment by sp5wwp** [comment]
Can the MCU transmit arbitrary baseband over FM now?

**Comment by silseva** [comment]
Yes, provided that you inject the signal externally, as the output stream driver is not implemented yet. The AT1846S driver can already support the transmission of digital modes.

**Comment by RE-Specto** [comment]
Hi! with [T-TWR v2.1](https://lilygo.cc/products/t-twr-rev2-1) released a while ago, I wonder if there's any plans for this small and promising platform? as far as I understand the development was waiting for the v2.1 release so the baseband signal routing through analog switches is available at the hardware level, which afaik now is the reality.

**Comment by RE-Specto** [comment]
thank you @silseva 
I really wish to find time.. ([still owe you](https://github.com/OpenRTX/OpenRTX/issues/275) a testing of the UV390 Plus 10w firmware as well :) and also have the v2.1 laying around on the shelf, as well as some [previous experience](https://github.com/RE-Specto/ESP32-NuteBarrels-Controller) with ESP32 (platformio with arduino library imported). but for now I'm massively overworked with my full time job, so it is hard to make any promises. nevertheless, I'm grateful to hear that there are plans for this platform, and hopeful to find time one day to give a helping hand with this one. I really believe that porting it to this platform will open a door to M17 for the masses, as it is finally an affordable off the shelf device to support the protocol, as well as a widely available micro-controller that many alternative custom designs can be delivered with.

**Comment by mrickey** [comment]
I know there isn't anyone working on this radio right now. But, could you add Persistence to the list. It would be nice if it remembered the VFO and call sign settings.

I have a 2.1 unit and would be happy to do some testing.

**Comment by silseva** [comment]
This is not stale. Currently there are issues with DMA transfers from the ADC, which are slowing down the development.

**Comment by RE-Specto** [comment]
thank you @silseva 
looks like the bot is pretty aggressive with setting the stale tag.

**Comment by silseva** [comment]
A bit, yes. Perhaps we should add a "long-development" tag, or something similar, to mark issues which seem to be stale but they are not.

**Comment by silseva** [comment]
I relaxed a bit the conditions of the close-stale job and added the "In progress" label: let's see if it works.

**Comment by chrisdebian** [comment]
The feature list here looks very comprehensive, is there anything still outstanding that isn't captured above? It would also be useful to know the current status of the ttwrplus branch relative to master: is there a plan or timeline for merging it in, or are there blockers preventing that? Given how much is already working, it would be good to understand what's needed to bring this to the same status as the other supported targets.

**Comment by silseva** [comment]
I know there are some issues with the mic sampling part, but edgetriggered and Niccolò are working on it.
It takes time, though...

**Comment by n1zzo** [comment]
We are working on enabling M17 and other digital modes on this platform, microphone sampling was demonstrated to work, now I am testing the emission of an M17 baseband.Currently there are no blocking points for merging this code into the main OpenRTX branch, but it’s worth to wait until we get full support of digital modes as well.If you have one T-TWR 2.1 and want to contibute join the Matrix or Discord channel dedicated to this radio and I can brief you.

Thanks!

**Comment by silseva** [comment]
> I relaxed a bit the conditions of the close-stale job and added the "In progress" label: let's see if it works.

It doesn't... I'll have a look in the next days.
